src/diffusers/pipelines/unidiffuser/modeling_text_decoder.py

```diff
@@ -0,0 +1,188 @@
+from typing import Optional
+
+import numpy as np
+import torch
+from torch import nn
+from transformers import GPT2LMHeadModel, GPT2Tokenizer
+
+from ...configuration_utils import ConfigMixin, register_to_config
+from ...models import ModelMixin
+
+
+# Modified from ClipCaptionModel in https://github.com/thu-ml/unidiffuser/blob/main/libs/caption_decoder.py
+class UniDiffuserTextDecoder(ModelMixin, ConfigMixin):
+    @register_to_config
+    def __init__(
+        self,
+        tokenizer: GPT2Tokenizer,
+        text_decoder: GPT2LMHeadModel,
+        prefix_length: int,
+        hidden_dim: Optional[int] = None,
+        use_hidden_dim: bool = True,
+    ):
+        """
+        Text decoder model for a image-text [UniDiffuser](https://arxiv.org/pdf/2303.06555.pdf) model. This is used to
+        generate text from the UniDiffuser image-text embedding.
+
+        Parameters:
+            tokenizer ([`GPT2Tokenizer`]):
+                Tokenizer of class
+                [GPT2Tokenizer](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2Tokenizer) for
+                the GPT-like text decoder model.
+            text_decoder ([`GPT2LMHeadModel`]):
+                Text decoder model of class
+                [GPT2LMHeadModel](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2LMHeadModel)
+                used to generate text from the UniDiffuser text embedding.
+            prefix_length (`int`):
+                TODO
+            hidden_dim (`int`, *optional*):
+                Hidden dim of the MLP if we encode the prefix.
+            use_hidden_dim (`bool`, *optional*, defaults to `True`):
+                Whether or not to use a MLP to encode the prefix.
+        """
+        super().__init__()
+        self.prefix_length = prefix_length
+
+        eos = "<|EOS|>"
+        special_tokens_dict = {"eos_token": eos}
+        self.tokenizer = tokenizer
+        self.tokenizer.add_special_tokens(special_tokens_dict)
+
+        self.transformer = text_decoder
+        # TODO: need to set the eos_token_id correctly
+        self.transformer.config.eos_token_id = self.tokenizer.eos_token_id
+        self.transformer.resize_token_embeddings(len(self.tokenizer))
```

Comment:
We can also make sure that the GPT2Transformer has the correct number of word embeddings before loading it so that we don't have to always resize the embedding every time at init